[feature] adding mypy for static typing

sergey-koldyba opened the pull request.

dmitrii-palisaderesearch commented on flake.nix:
```diff
@@ -24,6 +24,7 @@
             # Python
             uv
             ruff
+            python313Packages.mypy
```
I think in nixpkgs `mypy = python3Packages.mypy`, and `python3Packages = python3XXpackages, python3 = python3XX` where a concrete nixpkgs snapshot has a particular `python3XX` selected.

We currently don't pin a python3 version in template and let it roll with nixpkgs updates, therefore I think we should just specify `mypy` instead of pinning a Python version.

dmitrii-palisaderesearch commented on pyproject.toml:
```diff
@@ -14,6 +14,20 @@ dependencies = [
 ignore = ["*"]
 
+[tool.mypy]
+python_version = "3.13"
```
same here

dmitrii-palisaderesearch commented on .pre-commit-config.yaml:
```diff
@@ -27,7 +27,15 @@ repos:
+      - id: mypy
+        name: mypy
+        description: "Run 'mypy' for Python type checking"
+        entry: mypy . --python-executable ./.venv/bin/python
```
I think mypy will automatically use the Python in PATH, and it will be the correct one as long as we're in a direnv environment—so just `mypy .` should be enough?

sergey-koldyba replied:
Not really: https://github.com/PalisadeResearch/template/actions/runs/16519227725/job/46716998168
The reason is it redefines it's PATH un the start, explicitly setting nix's python path in front of venv ¯\_(ツ)_/¯ 

```bash
#! /nix/store/nravshvfviv75plf26nwfwzg8ldl3z8s-bash-5.2p37/bin/bash -e
PATH=${PATH:+':'$PATH':'}
PATH=${PATH/':''/nix/store/ssw8qs18c9scsya9h3jcbz1sxkfabcah-python3.13-mypy-1.15.0/bin'':'/':'}
PATH='/nix/store/ssw8qs18c9scsya9h3jcbz1sxkfabcah-python3.13-mypy-1.15.0/bin'$PATH
PATH=${PATH#':'}
PATH=${PATH%':'}
export PATH
PATH=${PATH:+':'$PATH':'}
PATH=${PATH/':''/nix/store/j1spz18w5a10r9nj4wpkyh0cx730rs4z-python3-3.13.4/bin'':'/':'}
PATH='/nix/store/j1spz18w5a10r9nj4wpkyh0cx730rs4z-python3-3.13.4/bin'$PATH
PATH=${PATH#':'}
PATH=${PATH%':'}
export PATH
export PYTHONNOUSERSITE='true'
exec -a "$0" "/nix/store/ssw8qs18c9scsya9h3jcbz1sxkfabcah-python3.13-mypy-1.15.0/bin/.mypy-wrapped"  "$@"
```

sergey-koldyba commented on .pre-commit-config.yaml:
```diff
@@ -27,7 +27,15 @@ repos:
+      - id: mypy
+        name: mypy
+        description: "Run 'mypy' for Python type checking"
+        entry: bash -c 'uv sync --extra dev --quiet && uv run mypy .'
```
Actually, after digging around I decided to move all to uv.
reason for that is possible library issues, unexpected paths, and weird magic like we discussed [here](https://github.com/PalisadeResearch/template/pull/29#discussion_r2230901562).

So let's install uv via nix and then let un + venv do the rest.

sergey-koldyba commented:
So moved to ty. Also, Switched language server to ty as well and removed Black Formatter extension(as we use ruff)
Didn't alter rules [defaults](https://docs.astral.sh/ty/reference/rules/) - they look fine to me.
